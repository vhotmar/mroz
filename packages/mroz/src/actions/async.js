import { createAction } from "redux-actions";

export const request = createAction("@@mroz/ASYNC_REQUEST", id => ({
  id
}));

export const success = createAction("@@mroz/ASYNC_SUCCESS", (id, data) => ({
  id,
  data
}));

export const fail = createAction("@@mroz/ASYNC_FAIL", (id, error) => ({
  id,
  error
}));
