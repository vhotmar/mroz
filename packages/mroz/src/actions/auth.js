import { createAction } from "redux-actions";

export const login = createAction("@@mroz/LOGIN", (token, profile) => ({
  token,
  profile
}));

export const logout = createAction("@@mroz/LOGOUT");
