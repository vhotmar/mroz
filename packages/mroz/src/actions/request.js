import { createAction } from "redux-actions";

export const request = createAction(
  "@@mroz/REQUEST_REQUEST",
  (id, key, { isPolling = false } = {}) => ({
    id,
    key,
    isPolling
  })
);

export const success = createAction(
  "@@mroz/REQUEST_SUCCESS",
  (id, key, data) => ({
    id,
    key,
    data,
    time: Date.now()
  })
);

export const fail = createAction("@@mroz/REQUEST_FAIL", (id, key, error) => ({
  id,
  key,
  error
}));

export const cancel = createAction("@@mrot/REQUEST_CANCEL", id => ({
  id
}));
