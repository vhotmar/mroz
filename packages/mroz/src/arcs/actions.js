import { compose } from "ramda";
import { Observable } from "rxjs";
import { addSuccess, addFail, addRequest } from "../epics/async";

const identity = x => x;

export default function actions(request, success, fail) {
  return compose(
    request ? addRequest(config => Observable.of(request(config))) : identity,
    success
      ? addSuccess((data, config) => Observable.of(success(data, config)))
      : identity,
    fail
      ? addFail((error, config) => Observable.of(fail(error, config)))
      : identity
  );
}
