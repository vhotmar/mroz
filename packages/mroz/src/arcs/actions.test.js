import actions from "./actions";
import { Observable } from "rxjs";
import asyncTest from "../epics/async.test.utils";

describe("actions arc", () => {
  it("should dispatch present actions", () => {
    asyncTest({
      config: actions(() => ({ type: "actions_request" })),
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });

    asyncTest({
      config: actions(undefined, () => ({ type: "actions_success" })),
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });

    asyncTest({
      work: () => Observable.throw("error"),
      config: actions(undefined, undefined, () => ({ type: "actions_fail" })),
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });
  });

  it("should dispatch all actions", () => {
    const config = actions(
      () => ({ type: "actions_request" }),
      () => ({ type: "actions_success" }),
      () => ({ type: "actions_fail" })
    );

    asyncTest({
      config,
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });

    asyncTest({
      work: () => Observable.throw("error"),
      config,
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });
  });
});
