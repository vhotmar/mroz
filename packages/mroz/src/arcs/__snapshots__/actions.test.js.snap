// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions arc should dispatch all actions 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "type": "actions_request",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "type": "actions_success",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`actions arc should dispatch all actions 2`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "type": "actions_request",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "type": "actions_fail",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
    },
    "type": "@@mroz/ASYNC_FAIL",
  },
]
`;

exports[`actions arc should dispatch present actions 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "type": "actions_request",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`actions arc should dispatch present actions 2`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "type": "actions_success",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`actions arc should dispatch present actions 3`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "type": "actions_fail",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
    },
    "type": "@@mroz/ASYNC_FAIL",
  },
]
`;
